hoodie-common/src/main/java/com/uber/hoodie/common/util/buffer/PushBasedBufferProducer.java

```diff
@@ -0,0 +1,66 @@
+/*
+ *  Copyright (c) 2017 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+
+package com.uber.hoodie.common.util.buffer;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.function.Function;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+/**
+ * Buffer multi-producer which allows custom functions to insert entries to buffer.
+ *
+ * @param <I> Type of entry produced for buffer
+ */
+public class PushBasedBufferProducer<I> extends BufferProducer<I> {
```

Comment:
The PullBased and PushBased names for these producers make it confusing (and concealing) to understand what the producer is actually doing, any specific advantage of separating them out like this ?

Reply:
These classes are intentionally generic. I needed flexibility in the Buffer implementation where I can dynamically compose a buffer with one or many "async" producers and a single "async" consumer with no assumption that the producer needs to only work with iterators and when to asynchronously start producers and consumers.  A single serial Producer that simply read records from iterator worked well before this change as that was the only use-case. But we also need to make 2 producers (Parquet and Log file) run in parallel. 

AbstractHoodieLogRecordScanner.scan simply append log records to this buffer (in unmerged reading).  As you can see, It does not read from an iterator unlike Parquet (we have an iterator for Parquet-Record Reader). 
I think PullBased (Iterator as first-class citizen, PushBased (opaque function producing single/batch of entries) and Composite (mix of Pull/Push) producers are good first-level producer types with which we can implement flexible rate/size controlled asynchronous multiProducer-singleConsumer models.

Reply:
Renamed the classes to IteratorBased and FunctionBased to make it explicit

Reply:
sounds good